components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/CertificateMgtDAO.java

```
@@ -169,34 +178,42 @@ public boolean addCertificate(String alias, String endpoint, int tenantId) throw
+                connection = APIMgtDBUtil.getConnection();
+                initialAutoCommit = connection.getAutoCommit();
+                connection.setAutoCommit(false);
+                preparedStatement = connection.prepareStatement(addCertQuery);
+                preparedStatement.setInt(1, tenantId);
+                preparedStatement.setString(2, endpoint);
+                preparedStatement.setString(3, alias);
+                result = preparedStatement.executeUpdate() == 1;
+                connection.commit();
+
+            } catch (SQLException e) {
+                try {
+                    if (connection != null) {
+                        connection.rollback();
+                    } else {
+                        log.error("Could not perform rollback since the connection is null.");
+                    }
+                } catch (SQLException e1) {
+                    log.error("Error while rolling back the transaction.", e1);
                 }
-            } catch (SQLException e1) {
-                log.error("Error while rolling back the transaction.", e1);
+                if (log.isDebugEnabled()) {
+                    log.debug("Error occurred while adding certificate metadata to database.", e);
+                }
+                handleException("Error while persisting certificate metadata.", e);
+            } finally {
+                APIMgtDBUtil.closeStatement(preparedStatement);
+                APIMgtDBUtil.closeAllConnections(preparedStatement, connection, null);
+                connection.setAutoCommit(initialAutoCommit);
             }
+        } catch (SQLException e) {
             if (log.isDebugEnabled()) {
-                log.debug("Error occurred while adding certificate metadata to database.", e);
+                log.debug("Error occurred in finally block while adding certificate metadata. ", e);
             }
-            handleException("Error while persisting certificate metadata.", e);
```

Exception swallowed?

---

fixed with https://github.com/wso2/carbon-apimgt/commit/3c8d1e61dc642ebe8179c483156798312a03382b